Transitory Sensitivity and Automatic Mechanism Analysis

mjohnson541 opened the pull request:
This PR adds functionality for transitory sensitivities and also the tools for automatic mechanism analysis.

hwpang commented on src/TransitorySensitivities.jl:
```diff
@@ -54,3 +54,64 @@ function splitnuminds(histo,n)
+end
+
+function getdividingtimescale(sim,t;taumax=1e6,taumin=1e-18,taures=10.0^0.5,usediag=true)
+    Jy = jacobiany(sim.sol(t),sim.domain.p,t,sim.domain,[]);
```
I feel that we should provide some way to feed in the interfaces to extend this to open system in the future. Not necessary to do in this PR

mjohnson541 replied:
Just left out the sim.interfaces XD

hwpang commented on src/TransitorySensitivities.jl:
```diff
@@ -465,3 +465,136 @@ function transitorysensitivitiesadjointexact(ssys::SystemSimulation,t,name;tau=N
+        gu = ReverseDiff.gradient(x->soladj.prob.f(x,soladj.prob.p,0.0)[ind],soladj.prob.u0)
+        dlm(lm,p,t) = dlmdt(lm,p,t,simadj.sol.prob.f,gu,soladj)
+        ode = ODEProblem(dlm,lm0,(tau,0.0))
+        lmsol = solve(ode,solveradj,reltol=reltoladj,abstola=abstoladj);
```
Typo, should be `abstol`?

mjohnson541 replied:
yep

hwpang commented on src/Domain.jl:
```diff
@@ -795,6 +795,28 @@ end
+        kfs .= d.p[length(d.phase.species)+1:end].*p[d.parameterindexes[1]-1+length(d.phase.species)+1:d.parameterindexes[1]-1+length(d.phase.species)+length(d.phase.reactions)]
+        Gs = d.p[1:length(d.phase.species)].+p[d.parameterindexes[1]-1+1:d.parameterindexes[1]-1+length(d.phase.species)]
+    end
+    krevs = getkfkrevs(d.phase,d.T,d.P,C,N,ns,Gs,d.diffusivity,V,0.0;kfs=kfs)[2]
```
Should we use `krevs .= ...` here?

mjohnson541 replied:
done

hwpang reviewed (changes requested):
@mjohnson541 Only several small request of changes. After addressing these, this PR can be merged!

mjohnson541 commented:
Okay I think all of those should be fixed. I also added FastGuassQuadrature to the dependencies and removed the manifest file (which probably shouldn't have been tracked by git in the first place).

hwpang commented:
This PR looks good to me! I will go ahead and merge it.
